File: ios/DanceFrame/PoseFrameProcessorPlugin.m
#if __has_include(<VisionCamera/FrameProcessorPlugin.h>)
#import <VisionCamera/FrameProcessorPlugin.h>
#else
#import "FrameProcessorPlugin.h"
#endif

#if __has_include(<VisionCamera/Frame.h>)
#import <VisionCamera/Frame.h>
#else
#import "Frame.h"
#endif

#import "PoseInferenceModule.h"

@interface PoseFrameProcessorPlugin : FrameProcessorPlugin
@end

@implementation PoseFrameProcessorPlugin

- (id _Nullable)callback:(Frame *)frame withArguments:(NSDictionary *_Nullable)arguments
{
  PoseInferenceModule *module = [PoseInferenceModule sharedInstance];
  if (module == nil) {
    return nil;
  }

  [module processFrame:frame withArguments:arguments];
  return nil;
}

VISION_EXPORT_FRAME_PROCESSOR(PoseFrameProcessorPlugin, poseFrameProcessor)

@end

File: android/app/src/main/java/com/danceframe/pose/PoseFrameProcessorPlugin.kt
package com.danceframe.pose

import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import com.mrousavy.camera.frameprocessors.FrameProcessorPluginRegistry

class PoseFrameProcessorPlugin : FrameProcessorPlugin() {
  override fun callback(frame: Frame, params: MutableMap<String, Any>?): Any? {
    PoseInferenceModule.getActiveInstance()?.processFrame(frame, params)
    return null
  }

  companion object {
    @Volatile private var isRegistered = false

    @JvmStatic
    fun register() {
      if (isRegistered) {
        return
      }

      FrameProcessorPluginRegistry.addFrameProcessorPlugin("poseFrameProcessor") { _, _ ->
        PoseFrameProcessorPlugin()
      }
      isRegistered = true
    }
  }
}
